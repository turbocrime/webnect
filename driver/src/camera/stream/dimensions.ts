import {
	Cam,
	CamFmtDepth,
	CamFmtInfrared,
	CamFmtVisible,
	CamRes,
} from "../enum.js";
import type { CamMode } from "../mode.js";

const visibleRes = {
	[CamRes.QVGA]: [320, 240],
	[CamRes.VGA]: [640, 480],
	[CamRes.SXGA]: [1280, 1024],
} as const;

const depthRes = {
	[CamRes.QVGA]: [320, 240],
	[CamRes.VGA]: [640, 480],
	[CamRes.SXGA]: [1280, 1024],
} as const;

const irRes = {
	[CamRes.QVGA]: [320, 240],
	[CamRes.VGA]: [640, 488],
	[CamRes.SXGA]: [1280, 1024],
} as const;

const visibleBpp = {
	[CamFmtVisible.BAYER_8B]: 8,
	[CamFmtVisible.YUV_16B]: 16,
} as const;

const depthBpp = {
	[CamFmtDepth.D_10B]: 10,
	[CamFmtDepth.D_11B]: 11,
} as const;

const irBpp = {
	[CamFmtInfrared.IR_10B]: 10,
} as const;

export const selectRes = <M extends CamMode>(mode: M) => {
	switch (mode?.stream) {
		case undefined:
			throw new RangeError("Invalid mode", { cause: mode });
		case Cam.VISIBLE:
			return visibleRes[mode.res] as M extends CamMode<Cam.VISIBLE>
				? (typeof visibleRes)[M["res"]]
				: never;
		case Cam.DEPTH:
			return depthRes[mode.res] as M extends CamMode<Cam.DEPTH>
				? (typeof depthRes)[M["res"]]
				: never;
		case Cam.INFRARED:
			return irRes[mode.res] as M extends CamMode<Cam.INFRARED>
				? (typeof irRes)[M["res"]]
				: never;
	}
};

export const selectBpp = <M extends CamMode>(mode: M) => {
	switch (mode?.stream) {
		case undefined:
			throw new RangeError("Invalid mode", { cause: mode });
		case Cam.VISIBLE:
			return visibleBpp[mode.format] as M extends CamMode<Cam.VISIBLE>
				? (typeof visibleBpp)[M["format"]]
				: never;
		case Cam.DEPTH:
			return depthBpp[mode.format] as M extends CamMode<Cam.DEPTH>
				? (typeof depthBpp)[M["format"]]
				: never;
		case Cam.INFRARED:
			return irBpp[mode.format] as M extends CamMode<Cam.INFRARED>
				? (typeof irBpp)[M["format"]]
				: never;
	}
};

export const selectFrameSize = (mode: CamMode) => {
	switch (mode?.stream) {
		case undefined:
			throw new RangeError("Invalid mode", { cause: mode });
		case Cam.VISIBLE: {
			const [width, height] = visibleRes[mode.res];
			const bpp = visibleBpp[mode.format];
			return width * height * (bpp / 8);
		}
		case Cam.DEPTH: {
			const [width, height] = depthRes[mode.res];
			const bpp = depthBpp[mode.format];
			return width * height * (bpp / 8);
		}
		case Cam.INFRARED: {
			const [width, height] = irRes[mode.res];
			const bpp = irBpp[mode.format];
			return width * height * (bpp / 8);
		}
	}
};
