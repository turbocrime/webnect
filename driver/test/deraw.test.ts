import { beforeEach, describe, expect, test } from "vitest";
import {
	bayerToRgbaFloat16,
	unpackGrayToRgbaFloat16,
	yuvToRgbaFloat16,
} from "../src/camera/deraw/formats.js";
import {
	MODE_DEPTH,
	MODE_INFRARED,
	MODE_VISIBLE_BAYER,
	MODE_VISIBLE_YUV,
} from "../src/camera/mode.js";
import { selectRes } from "../src/camera/stream/dimensions.js";

describe("deraw", () => {
	let canvasContext: CanvasRenderingContext2D;

	beforeEach((ctx) => {
		const testDiv = document.createElement("div");
		testDiv.id = ctx.task.id;
		testDiv.textContent = ctx.task.name;
		document.body.appendChild(testDiv);

		const canvas = document.createElement("canvas");
		testDiv.appendChild(canvas);

		canvas.width = 640;
		canvas.height = 488;

		// biome-ignore lint/style/noNonNullAssertion: canvas should have a context
		canvasContext = canvas.getContext("2d")!;
	});

	test("depth 11bit", async () => {
		const [width, height] = selectRes(MODE_DEPTH);

		const rawBuffer = await fetch(
			new URL("./data/depth.11bit.raw", import.meta.url),
		).then((response) => response.arrayBuffer());

		const rgbaBuffer = unpackGrayToRgbaFloat16(11, rawBuffer);

		const imageData = new ImageData(
			new Float16Array(rgbaBuffer) as never,
			width,
			height,
			{ pixelFormat: "rgba-float16" } as never,
		);
		canvasContext.putImageData(imageData, 0, 0);

		const hashCanvas = await crypto.subtle.digest(
			"SHA-256",
			canvasContext.getImageData(0, 0, width, height).data,
		);
		expect(hashCanvas).toMatchSnapshot();
	});

	test("infrared 10bit", async () => {
		const [width, height] = selectRes(MODE_INFRARED);

		const rawBuffer = await fetch(
			new URL("./data/infrared.10bit.raw", import.meta.url),
		).then((response) => response.arrayBuffer());

		const rgbaBuffer = unpackGrayToRgbaFloat16(10, rawBuffer);

		const imageData = new ImageData(
			new Float16Array(rgbaBuffer) as never,
			width,
			height,
			{ pixelFormat: "rgba-float16" } as never,
		);
		canvasContext.putImageData(imageData, 0, 0);

		const hashCanvas = await crypto.subtle.digest(
			"SHA-256",
			canvasContext.getImageData(0, 0, width, height).data,
		);
		expect(hashCanvas).toMatchSnapshot();
	});

	test("yuv", async () => {
		const [width, height] = selectRes(MODE_VISIBLE_YUV);

		const rawBuffer = await fetch(
			new URL("./data/yuv.raw", import.meta.url),
		).then((response) => response.arrayBuffer());

		const rgbaBuffer = yuvToRgbaFloat16(rawBuffer);

		const imageData = new ImageData(
			new Float16Array(rgbaBuffer) as never,
			width,
			height,
			{ pixelFormat: "rgba-float16" } as never,
		);
		canvasContext.putImageData(imageData, 0, 0);

		const hashCanvas = await crypto.subtle.digest(
			"SHA-256",
			canvasContext.getImageData(0, 0, width, height).data,
		);
		expect(hashCanvas).toMatchSnapshot();
	});

	test("bayer", async () => {
		const [width, height] = selectRes(MODE_VISIBLE_BAYER);

		const rawBuffer = await fetch(
			new URL("./data/bayer.raw", import.meta.url),
		).then((response) => response.arrayBuffer());

		const rgbaBuffer = bayerToRgbaFloat16(640, rawBuffer);

		const imageData = new ImageData(
			new Float16Array(rgbaBuffer) as never,
			width,
			height,
			{ pixelFormat: "rgba-float16" } as never,
		);
		canvasContext.putImageData(imageData, 0, 0);

		const hashCanvas = await crypto.subtle.digest(
			"SHA-256",
			canvasContext.getImageData(0, 0, width, height).data,
		);
		expect(hashCanvas).toMatchSnapshot();
	});
});
