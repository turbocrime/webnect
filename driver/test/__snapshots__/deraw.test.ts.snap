// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`deraw > bayer 1`] = `
ArrayBuffer [
  8,
  -51,
  7,
  -93,
  -88,
  -30,
  -20,
  -2,
  -40,
  42,
  1,
  -80,
  -19,
  95,
  -75,
  -68,
  22,
  -27,
  -58,
  -25,
  48,
  122,
  -105,
  -110,
  -82,
  -123,
  67,
  -41,
  -127,
  -76,
  82,
  -7,
]
`;

exports[`deraw > depth 11bit 1`] = `
ArrayBuffer [
  117,
  -52,
  68,
  51,
  -49,
  119,
  121,
  29,
  32,
  9,
  29,
  -87,
  49,
  69,
  -103,
  79,
  -115,
  75,
  -96,
  -20,
  60,
  -10,
  97,
  14,
  23,
  -78,
  -53,
  9,
  21,
  43,
  -33,
  89,
]
`;

exports[`deraw > infrared 10bit 1`] = `
ArrayBuffer [
  -105,
  -1,
  -126,
  48,
  -91,
  37,
  -49,
  113,
  57,
  55,
  58,
  111,
  -76,
  -99,
  1,
  -83,
  94,
  -86,
  29,
  112,
  -28,
  22,
  -123,
  -30,
  58,
  29,
  -11,
  -83,
  57,
  103,
  87,
  -38,
]
`;

exports[`deraw > yuv 1`] = `
ArrayBuffer [
  -123,
  -77,
  -68,
  -33,
  -81,
  121,
  -89,
  43,
  83,
  12,
  27,
  -16,
  5,
  46,
  27,
  -58,
  61,
  1,
  -86,
  -15,
  70,
  88,
  84,
  64,
  14,
  -75,
  53,
  59,
  -1,
  -28,
  -64,
  54,
]
`;
